### ISSUE TITLE
make_hanging_node_constraints() with p::d::Triangulation + hp::DoFHandler Error

### ISSUE THREAD

**Issue by drwells** [dealii/dealii#9517]
This is from the mailing list - Chaitanya Dev is using p::d::Triangulation with hp::DoFHandler and ran into problems. We end up violating `cell->face(face)->child(c)->n_active_fe_indices() == 1`: I didn't look closely but this seems like a bug.

Here is a slightly edited version of the original message:

I am trying to use parallel::distributed::Triangulation with hp::DoFHandler. I run into the below error in DoFTools::make_hanging_node_constraints.

I work on shape optimization using embedding domain discretization techniques. In my problems, I have to repeatedly refine the mesh and approximate the geometry. Then I assign different FESystem for the cells within and outside the geometry. 

I have implemented the code in serial with serial triangulation and hp::DoFHandler and the program works. 

Now I wish to solve larger problems so I have decided to parallelize the code and hence I use parallel::distributed::Triangulation.

Here I run into a problem in system_setup, specifically in the function DoFTools::make_hanging_node_constraints.

Hear is the error message.
```
An error occurred in line <1119> of file </spack-src/source/dofs/dof_tools_constraints.cc> in function
    void dealii::DoFTools::internal::make_hp_hanging_node_constraints(const DoFHandlerType&, dealii::AffineConstraints<number>&) [with DoFHandlerType = dealii::hp::DoFHandler<2, 2>; number = double]
The violated condition was: 
    cell->face(face)->child(c)->n_active_fe_indices() == 1
Additional information: 

Stacktrace:

#0  /opt/dealii-9.1.1/lib/libdeal_II.g.so.9.1.1: void dealii::DoFTools::internal::make_hp_hanging_node_constraints<dealii::hp::DoFHandler<2, 2>, double>(dealii::hp::DoFHandler<2, 2> const&, dealii::AffineConstraints<double>&)
#1  /opt/dealii-9.1.1/lib/libdeal_II.g.so.9.1.1: void dealii::DoFTools::make_hanging_node_constraints<dealii::hp::DoFHandler<2, 2>, double>(dealii::hp::DoFHandler<2, 2> const&, dealii::AffineConstraints<double>&)
#2  hp-dist-test: TestProblem<2>::setup_system()
#3  hp-dist-test: TestProblem<2>::run()
#4  hp-dist-test: main
```

This is my setup_system function.
```cpp
template<int dim>
void TestProblem<dim>::setup_system() {

	dof_handler.initialize(triangulation, fe_collection);

	for (auto cell : dof_handler.active_cell_iterators()) {
		if (!cell->is_locally_owned())
			continue;
		if (cell->user_flag_set())
			cell->set_active_fe_index(0);
		else
			cell->set_active_fe_index(1);
	}

	dof_handler.distribute_dofs(fe_collection);
	locally_owned_dofs = dof_handler.locally_owned_dofs();
	DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);

	hanging_node_constraints.clear();
	hanging_node_constraints.reinit(locally_relevant_dofs);

	DoFTools::make_hanging_node_constraints(dof_handler,
			hanging_node_constraints);
	hanging_node_constraints.close();
}
```
I did some debugging and found that the Assert fail cell->face(face)->child(c)->n_active_fe_indices() == 1 is failing on the ghost cells. To avoid this error  I tried to set_active_fe_index for the ghost cells, but this is not allowed. 

Please let me know where I am going wrong.

I have attached a MWE. You can reproduce the error with mpirun for np >= 2 cores (the code will work for np=1, because no ghost cells). 

Thank you,
Chaitanya Dev.

**Comment by drwells** [comment]
[CMakeLists.txt](https://github.com/dealii/dealii/files/4188066/CMakeLists.txt)
[hp-dist-test.txt](https://github.com/dealii/dealii/files/4188069/hp-dist-test.txt)

**Comment by marcfehling** [comment]
This should've been fixed in #8365, which is not included in `deal.II-9.1.1`.

Compiling from the master branch should do the trick.

**Comment by drwells** [comment]
You're right - I confirmed that this program executes without assertion failures in debug mode on `master`.
